Marlin/src/HAL/STM32/u8g_com_stm32duino_ssd_i2c.cpp

```diff
@@ -0,0 +1,166 @@
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ *
+ */
+
+
+/**
+ * 2 wire I2C COM driver
+ *
+ * Handles both hardware I2C and software I2C.
+ *
+ * The WIRE library is used for the hardware I2C.  That means that
+ * ANY hardware SDA and SLC pairs can be selected and the WIRE library
+ * will automatically enable the correct hardware.
+ *
+ * The SoftWire library is used for the software I2C. That means that
+ * ANY pin can be selected for SDA and SLC.
+ *
+ * This driver requires the SDA and SLC pins be named:
+ *   DOGLCD_SDA
+ *   DOGLCD_SCL
+ *
+ * This allows independence from other I2C devices (mostly EEPROMs) that
+ * usually use I2C_SDA_PIN and I2C_SLC_PIN.
+ *
+ * The following MUST be present in the platformio environment used to
+ * compile Marlin. This required because
+ *
+ * The flag LCD_I2C_SOFT is used to select which I2C library is used.
+ * It must be enabled in the platformio environment used to compile
+ * Marlin.  Enabling it anywhere else results in LCD_I2C_SOFT not being
+ * enabled during crucial portions of the compile process which results
+ * in defaulting to hardwired operation.  The net result is the following
+ * MUST be in the platformio environment in order to use the software I2C
+ * system:
+ *    build_flags = -DLCD_I2C_SOFT
+ *    lib_deps    = stevemarple/SoftWire@^2.0.9
+ *                  stevemarple/AsyncDelay@^1.1.2
+ *
+ */
+
+#include "../../inc/MarlinConfigPre.h"
+
+#if (defined(ARDUINO_ARCH_STM32) && (defined(U8GLIB_SH1106) || defined(IS_U8GLIB_SSD1306) || defined(U8GLIB_SSD1309)))
+
+#include <U8glib-HAL.h>
+
+#include "../../MarlinCore.h"  // so can get SDA & SCL pins
+
+/*
+  BUFFER_LENGTH is defined in libraries\Wire\utility\WireBase.h
+  Default value is 32
+  Increate this value to 144 to send U8G_COM_MSG_WRITE_SEQ in single block
+*/
+
+#if !defined(BUFFER_LENGTH) || BUFFER_LENGTH >= 144
+  #define BUFFER_LENGTH 32
+#endif
+#define I2C_MAX_LENGTH (BUFFER_LENGTH - 1)
+
+#ifdef LCD_I2C_SOFT
+  #include <SoftWire.h>
+  #include <AsyncDelay.h>
+
+  char swTxBuffer[BUFFER_LENGTH];
+  char swRxBuffer[BUFFER_LENGTH];
+  SoftWire sw(DOGLCD_SDA, DOGLCD_SCL);
+  #define I2C_ITF sw
+
+#elif !defined(HAL_I2C_MODULE_DISABLED)
+  #include <Wire.h>
+  #define I2C_ITF Wire
+  #ifndef MASTER_ADDRESS
+    #define MASTER_ADDRESS 0x01
+  #endif
+#else
+  #error "unsupported I2C configuration"
+#endif
+
+
+static uint8_t control;
+static uint8_t msgInitCount = 2; // Ignore all messages until 2nd U8G_COM_MSG_INIT
+
+uint8_t u8g_com_stm32duino_ssd_i2c_fn(u8g_t *u8g, uint8_t msg, uint8_t arg_val, void *arg_ptr)
+{
+  if (msgInitCount) {
+    if (msg == U8G_COM_MSG_INIT) msgInitCount--;
+    if (msgInitCount) return -1;
+  }
+
+  switch (msg)
+  {
+    case U8G_COM_MSG_INIT:
+      #ifdef LCD_I2C_SOFT
+        I2C_ITF.setClock(100000);
```

Comment:
Have you considered maxing out clock speed by default?
This out results in "only" 138kHz in an SKR3 (550MHz).

Reply:
Done but it didn't seem to affect anything. I've set both to 400K but both actuals come in at about 100K.  I expected to get more speed on the hardware driver version.   I'll definitely investigate.